addOns/pscanrules/src/main/java/org/zaproxy/zap/extension/pscanrules/SameOriginMethodExecutionScanRule.java

```diff
@@ -0,0 +1,227 @@
+            // which is likely to be the case in JSONP responses.
+        }
+
+        for (HtmlParameter param : allUrlParameters) {
+            Matcher funcMatcher = JSONP_PATTERN.matcher(respBody);
+            while (funcMatcher.find()) {
+                String matchedFuncName = funcMatcher.group(1);
+                if (matchedFuncName.equals(param.getValue())) {
+                    LOGGER.info(
+                            "[ALERT] Parameter '" + param.getName() + "' is vulnerable to SOME.");
+                    createAlert(
+                                    "jsonp",
+                                    (sfdHeader != null && sfdHeader.equals("script"))
+                                            ? Alert.CONFIDENCE_HIGH
+                                            : Alert.CONFIDENCE_MEDIUM,
+                                    param.getValue(),
+                                    refererHeader)
+                            .raise();
+                }
+            }
+        }
+    }
+
+    private AlertBuilder createAlert(
+            String type, int confidence, String evidence, String otherInfoParam) {
+        return newAlert()
+                .setRisk(Alert.RISK_HIGH)
+                .setConfidence(confidence)
+                .setDescription(Constant.messages.getString(MESSAGE_PREFIX + "desc"))
+                .setOtherInfo(
+                        Constant.messages.getString(
+                                MESSAGE_PREFIX + type + ".extrainfo", otherInfoParam))
+                .setEvidence(evidence)
+                .setSolution(Constant.messages.getString(MESSAGE_PREFIX + "soln"))
+                .setCweId(359) // CWE-359: Exposure of Private Information ('Privacy Violation')
+                .setWascId(13); // WASC-13: Information Leakage
+    }
+
+    private void performHeuristicsCheck(
+            HtmlParameter param, List<Element> scriptTags, boolean isSwf) {
+
+        LOGGER.info(param.toString());
+        if (CALLBACK_PARAMS.contains(param.getName()) && isSwf) {
+            createAlert("swf", Alert.CONFIDENCE_MEDIUM, param.getName(), param.getName());
+        }
+
+        for (Element tag : scriptTags) {
+            String src = tag.getAttributeValue("src");
+            LOGGER.info(src);
+            for (String callbackParam : CALLBACK_PARAMS) {
+                String expectedUrlParam = callbackParam + "=" + param.getValue();
+                LOGGER.info(expectedUrlParam);
+
+                if (src.contains(expectedUrlParam)) {
+                    createAlert(
+                                    "heuristics",
+                                    Alert.CONFIDENCE_MEDIUM,
+                                    expectedUrlParam,
+                                    param.getName())
+                            .raise();
+                }
+            }
+        }
+    }
+
+    private TreeSet<HtmlParameter> getUrlParameters(HttpMessage newReq) {
```

Comment:
Why not use getUrlParams directly?

Reply:
I was doing that initially, but in the application some endpoints were sending the callback parameter within another parameter after URL encoding it, so that is why I am decoding it and then chaning the URL in the HttpMessage object and then fetching the URL params. I think the LinkedIn team realised this as well because they have done the same thing:

![image](https://github.com/zaproxy/zap-extensions/assets/59091280/f2acfbdb-ff8a-4383-8d1f-4d7994adeb1c)

Example: https://public-firing-range.appspot.com/reverseclickjacking/multipage/ParameterInQuery/OtherParameter/WithoutXFO/?q=%26callback%3Dfoo%23